greenpeace/planet4-plugin-gutenberg-blocks
# PLANET-4808: Convert Split Two Columns to WYSIWYG

Ref: https://jira.greenpeace.org/browse/PLANET-4808
Also solves:
- https://jira.greenpeace.org/browse/PLANET-4317
- https://jira.greenpeace.org/browse/PLANET-5205

## Conversion of Split Two Columns block to WYSIWYG.

<img width="600" alt="Screenshot 2020-08-10 at 09 27 18" src="https://user-images.githubusercontent.com/617346/89760174-16549380-daec-11ea-9da8-37e142ec9fa5.png">


Just to be clear:
![15-i-have-no-idea jpg=w768](https://user-images.githubusercontent.com/617346/89758737-c7592f00-dae8-11ea-80c2-4861eaa12d97.jpeg)
So don't hesitate to make me redo everything.


## Details

The port to wysiwyg is pretty straightforward, `SplittwocolumnsFrontend.js` displays the frontend version, `SplittwocolumnsInPlaceEdit.js` is the same block in edit mode and `SplittwocolumnsSettings.js` has the sidebar settings.  
What's more complicated is the handling of old data. Technical choices for this are specified in the **Additions** section.    
Some state has to be saved, so that we know when to overwrite content with the selected issue/tag content, and when the user has edited something (in which case we keep the user data). The `edited` property keeps track of this.

## Tests

- Create, edit and view a new block Split Two Columns in a new page.
  - The only text not editable in-place is the Tag name
- View and edit existing blocks to verify they still work as intended.

## Other fixes

Fix for **PLANET-5205** Implement data attributes on Split Two Columns Block

Fix for **PLANET-4317** Split 2 column block: Text in button has a line break

## Additions

### Addition of `.babelrc` file to allow for more js syntaxes
It makes the code using API responses more concise.

- nullish coalescing operator `??` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator
- optional chaining operator `?.` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

### DIsplay of CMB2 fields in the REST API response

Introduced by https://github.com/greenpeace/planet4-master-theme/pull/1162, it was necessary to gather the information that was previously fetched by the PHP code (overwrite of title and description on the selected issue). CMB2 fields are not in the REST response by default.

### Display of `tag_attachment_id` in the REST API response

I needed the ID of the image associated to a tag, so it now appears when this block is registered. See `Splittwocolumns::__construct()`.

### New endpoint `POST planet4/v1/update_block/<block_name>`

Receives the current block structure, and returns the same block in the last structure version.

The reasoning is the following:  
- The block needs a new structure if we want to use React instead of Twig on the frontend. Twig was doing server-side rendering, which is cacheable and doesn't impact visitors as much as frontend side rendering.
- Either we convert existing blocks through a patch or on the fly before display, there has to be a PHP function that does this job, so it stays fast and compatible with the previous version of this block. Specifically, _images, descriptions, titles, links_, have to be fetched using only the issue, tag and images ids.  
- We could add a JS `migrate` function in the block code that does all this, but it would add many requests on the front side for visitors, and subsequently more complexity for the frontend code, so it has to be done on the backend.
- The new version adds these parameters on edition, they are available on the frontend without any more work. It doesn't need compatibility code.
- If we don't patch existing blocks, for better usability and less version juggling, it is easier to convert old blocks to the new structure during edition of those blocks; because the code for this is already ready for display compatibility in `SplitTwoColumns::update_data()`, we can also call this code through the API to transform the block, instead of writing this again in JS.

### Versioning of the structure

For future use, I considered that adding a version number to the block data was an easier way to handle changes that trying to compare properties existence/content.
